### DISCUSSION TITLE
Subscribe to DB updates done by different session

### DISCUSSION THREAD

**Discussion by ddenev** [appwrite/appwrite#3310] [Q&A]
Is it possible to subscribe to Realtime collections events but on the conditions they have been initiated by a particular user session?

My use case is: I have a web app and a mobile app - both connecting to the same Appwrite project and both able to update the database. I want when I update the DB from the web app, the mobile app to automatically sync it's local state from the database. And vise versa. So I need to be able to subscribe to DB updates but only if they were done by another session, not the current one - otherwise I'll be fetching data on the same app on every DB update.

Is this currently possible or are there plans to implement this in Appwrite?

**Comment by ddenev** [comment]
@eldadfux , sorry to tag you personally on this but I have the feeling the discussions are not much read by the core team - I have asked 2 questions and never got an Appwrite team member answering.
What's the most efficient way of reaching to the team?

Thank you and best regards!

**Reply by gewenyu99** [reply]
👀 Just saw this. The best way to get noticed is probably on Discord: appwrite.io/discord. That said, I should probably check discussions more often!

With that said, if you use real-time, you should try using a collection (or sever) as your ViewModel, and update what you render on the frontend only when you receive the update from real-time.

Here's the logic: Render view from collection data -> user clicks button -> app changes collection -> collection update triggers realtime event -> use realtime event to trigger update to rendered view

You only re-render after getting the realtime update back from Appwrite. This solves consistency issue across multiple apps. Optimistically updating the local state before getting a realtime update back is also risky if the request fails to update the collection, then you have to revert the local state anyway.

LMK if this answers your question. 

Firestore like local-state syncing is something frequently requested, and something we're exploring, but I don't believe it's planned just yet.

**Reply by ddenev** [reply]
@gewenyu99 , thank you very much for the reply!

That was the approach that I considered initially but abandoned after some time b/c Realtime subscriptions are not granular enough to give me the level of control I need.

In fact I am doing a similar thing (in terms of using the DB as my ViewModel) - only I manually issue fetches from the DB whenever I finish my updates to the DB.

Let me give you an example: imagine a collection that has an `order` attribute - to represent the ordering of the collection items when I represent them in the UI. I can reorder the items, which would require updating the `order` attribute for multiple documents.
Imagine I want to insert an item in this collection. That would require:

1. calling `createDocument()` to insert the new item in the collection
2. calling `updateDocument()` for every other document in the collection - to update the `order` attribute
With Realtime that would fire a bunch of events that I wouldn't care about until all of the above has finished. When I do it manually, I have the fine level of control **when** to fetch from the DB.

Another example - I have a collection named "Categories" and it contains documents in some form of a parent-child relationship, using `parent_id` and `children_ids` attributes. So, whenever I insert a document, I also need to do a bunch of updates to other documents in order to maintain the relationships. And again - that would fire a whole bunch of Realtime events in which I am not interested, I'm only interested after all inserts/updates had finished.

This are just 2 simple examples, I have (and will have) even more complex cases but all boil down to: I need to do a bunch of creates/deletes/updates to the DB before I fetch my collections in order to update the rendering.

Which brings me to the question (also inspired by what you described above): when I update the DB from my web app, if I want my mobile app to be in sync, it will still have to rely on Realtime subscriptions - so how do I overcome the "multitude of events" in my mobile app each time I insert a Category document in my web app?
This kinda negates my original question but thanks to your feedback I had to think deeper about this.

Maybe Realtime is just not ready yet for such use cases... or maybe I'm missing something really obvious

**Comment by ddenev** [comment]
I though some more on the use case and the problem and realized that the way Realtime subscriptions currently work, they will not be able to help me solve my problem.

Let me explain.

Let's consider the example of a collection with the `order` attribute (see above) where I need to update all other documents (update their `order`) before I insert a new document. So, one atomic "create document" business function is actually many `updateDocument()` plus one `createDocument()`.
Initially I thought - OK, no problem - I will subscribe to `collections.[ID].documents` and will update my local state based on the received events from Realtime:

- The multiple `collections.[ID].documents.*.update` events, each one will trigger updating one of the items in the local state list
- The last `collections.[ID].documents.*.create` event will trigger inserting a new item in my local state list

and my state would be consistent and in sync. Right?

Well, no! The thing is all of the updates (and the create) would trigger a re-render of my UI. Why? Because they come asynchronously, so each update event from Realtime will asynchronously update my state, which in turn will trigger re-rendering. Same for the create event.
Compare that to if I directly manipulate my state - I would have the same business logic for inserting an item (multiple updates + insert) but this time these would happen synchronously, so there will be only one render at the end (I use Vue but same would apply to React).

Therefore bluntly subscribing to create/update/delete events will not help me keep my state in a consistent state and avoid multiple re-renders. I need a way to 'instruct' Realtime that I have finished with my business function (which could include multiple updates/creates/deletes to the backend) so that now it fires notifications to everyone interested (my web app or my mobile app) that they should fetch data and update their state.

You sure understand well that in any business application it is very rare that a single "create" business operation is only a single "insert" of a single document into a single collection in the DB. Most often creating an object requires several backend-related operations - that's one reason why DBs have transactions - to encapsulate multiple operations into an atomic one. So maybe what I need is Realtime to allow me to subscribe to "transactions", instead of only to a basic set of CUD events?

Anyone know how is this possible to achieve with Appwrite?

PS: @gewenyu99 , can we keep the discussion here as IMHO Discord is good for quick questions/answers but fails to adequately track deeper/longer discussions?

**Reply by stnguyen90** [reply]
Appwrite does [want to support transactions](https://github.com/appwrite/appwrite/issues/2788). 

For your specific use case, syncing the order of things, I would recommend changing the way you use order such that you can insert something between 2 other items without having to update other documents.

**Reply by ddenev** [reply]
Thanks for replying, @stnguyen90 

The `order` use-case was just one of the many examples to illustrate that there are many cases when one needs to perform several DB operations in order to execute one business operation. So just resolving the `order` use-case would not automatically help resolving all the other use-cases :)

> I would recommend changing the way you use order such that you can insert something between 2 other items without having to update other documents.

Interesting. Any idea how can I do that? How can I insert a document into DB and preserve ordering without affecting other documents?

**Reply by gewenyu99** [reply]
You can filter the event you receive by looking at the list of `events` returned so you only trigger UI changes when you receive a "create" event. 

With that said, updating every single document in a collection on insert to update order is usually considered an antipattern. There are many interesting ways to resolve this problem. All of them I know of require a periodic background task to clean up.

For example, on insertion, if you wish to insert between records `A` and `B`, you can insert `AA` then `AAA` then `AB`, the results of which will be like this:
```
‘A’
‘AA’
‘AAA’
‘AB’
‘B’
```
Then, periodically, we can flatten the hierarchy again and return to:
```
`A`
`B`
`C`
... etc
```

Some DBs use a paging system, where ordered items are organized into pages, searches first find the page, then find the particular place on the page where the item is stored. When inserting or deleting items, you only worry about the ordering within the page. Periodically, pages are reorganized through a background task to keep things in order. This is most useful for sparse insertions on large catalogs.

Anyway, I got sidetracked. We do plan to support transactions, it's a must. I have also raised this when I joined and I'll continue to remind the team of this.

Most DB that need to be strongly consistent and ordered AND handle large amounts of data will do this to some capacity, anyway. 

For now, you can actually create "batch execution" of complex business logic as a Cloud Function and subscribe to `functions.*.executions.*.update` and filter for specific functions, and filter for status to be execution complete. You only query the DB once the function finishes, instead of waiting for a DB update.

What you can try as well, is to have your view as a "materialized view" of sorts, where several DBs compose a view, when they're updated, they can trigger another cloud function to compile them and update the materialized view, and have your frontend subscribe to the materialized view.

These are ideas, while the design is about lots of trade-offs. 

Here's what I take away and will share with the team. 
1. We need transactions
2. We need something like Firestore's syncing functionality to help manage state consistency in real-time.

I hope I was able to help, or at least provided an interesting conversation! Your use case is not unique, we've heard similar requests, and is something we'll try to better support in the future.

Cheers!!!

**Reply by ddenev** [reply]
Thanks, @gewenyu99 !

> You can filter the event you receive by looking at the list of events returned so you only trigger UI changes when you receive a "create" event.

That would mean that now I'm forced to always execute `create` last but that will not always be possible/desirable as in some cases I need to `create` first (for example I need the new ID).

Thanks for the ideas for ordering but again - ordering was just one example to illustrate the use case where we need to execute several actions atomically and only when they finish (all of them) we are now interested in being notified.

> For now, you can actually create "batch execution" of complex business logic as a Cloud Function and subscribe to functions.*.executions.*.update and filter for specific functions, and filter for status to be execution complete. You only query the DB once the function finishes, instead of waiting for a DB update.

Now that's an interesting idea. I also thought of something similar - to have a separate collection where I would add a record each time my atomic business operation finishes (the record will carry info about the operation). And then I would subscribe to that collection only, thus being able to control exactly when to sync and exactly what I want to sync to the client.

> Here's what I take away and will share with the team.
> 1. We need transactions
> 2. We need something like Firestore's syncing functionality to help manage state consistency in real-time.
>
> I hope I was able to help, or at least provided an interesting conversation! Your use case is not unique, we've heard similar requests, and is something we'll try to better support in the future.

Thank you for "carrying the torch" and for the detailed feedback and ideas! It's encouraging to know that this is being thought about in the Appwrite team :)
